File: src/main/java/org/wysko/midis2jam2/instrument/percussion/Tom.java
package org.wysko.midis2jam2.instrument.percussion;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;

import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;

public class Tom extends StickDrum {
	public Tom(Midis2jam2 context, List<MidiNoteOnEvent> hits, TomPitch pitch) {
		super(context, hits);
		drum = context.loadModel("DrumSet_Tom.obj", "DrumShell.bmp", Midis2jam2.MatType.UNSHADED);
		drum.setLocalScale(pitch.scale);
		recoilNode.attachChild(drum);
		recoilNode.attachChild(stickNode);
		highLevelNode.attachChild(recoilNode);
		highLevelNode.setLocalTranslation(pitch.location);
		highLevelNode.setLocalRotation(pitch.rotation);
		
		
		if (pitch == TomPitch.HIGH_FLOOR || pitch == TomPitch.LOW_FLOOR) {
			stickNode.setLocalRotation(new Quaternion().fromAngles(0, rad(80), 0));
			stickNode.setLocalTranslation(10, 0, 0);
		} else {
			stickNode.setLocalTranslation(0, 0, 10);
		}
		
		stick.setLocalRotation(new Quaternion().fromAngles(rad(MAX_ANGLE), 0, 0));
	}
	
	@Override
	public void tick(double time, float delta) {
		drumRecoil(time, delta);
		handleStick(time, delta);
	}
	
	public enum TomPitch {
		LOW_FLOOR(
				new Vector3f(1.5f, 1.5f, 1.5f),
				new Vector3f(20, 20, -60),
				new Quaternion().fromAngles(rad(-2), rad(180), rad(-10))),
		
		HIGH_FLOOR(new Vector3f(1.4f, 1.4f, 1.4f),
				new Vector3f(17, 21, -75),
				new Quaternion().fromAngles(rad(-5), rad(180), rad(-15))),
		
		LOW(new Vector3f(1.2f, 1.2f, 1.2f),
				new Vector3f(10f, 27 + 2, -82),
				new Quaternion().fromAngles(rad(60), rad(-30), 0)),
		
		LOW_MID(new Vector3f(1f, 1f, 1f),
				new Vector3f(0, 30 + 2, -85),
				new Quaternion().fromAngles(rad(60), 0, 0)),
		
		HIGH_MID(new Vector3f(0.8f, 0.8f, 0.8f),
				new Vector3f(-9, 29 + 2, -82),
				new Quaternion().fromAngles(rad(60), rad(20), 0)),
		
		HIGH(new Vector3f(0.6f, 0.6f, 0.6f),
				new Vector3f(-15, 27 + 2, -78),
				new Quaternion().fromAngles(rad(50), rad(40), 0));
		
		final Vector3f scale;
		final Vector3f location;
		final Quaternion rotation;
		
		TomPitch(Vector3f scale, Vector3f location, Quaternion rotation) {
			this.scale = scale;
			this.location = location;
			this.rotation = rotation;
		}
	}
}

File: src/main/kotlin/org/wysko/midis2jam2/instrument/family/percussion/drumset/kit/Tom.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion.drumset.kit

import com.jme3.math.Quaternion
import com.jme3.math.Vector3f
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import org.wysko.kmidi.midi.event.NoteEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.StickType
import org.wysko.midis2jam2.instrument.algorithmic.Striker
import org.wysko.midis2jam2.instrument.family.percussion.drumset.DrumSetInstrument
import org.wysko.midis2jam2.util.Utils
import org.wysko.midis2jam2.util.Utils.rad
import org.wysko.midis2jam2.world.modelD
import kotlin.time.Duration

private val STICK_NODE_OFFSET = Vector3f(0f, 0f, 10f)

/** A Tom. */
class Tom(context: Midis2jam2, hits: MutableList<NoteEvent.NoteOn>, pitch: TomPitch, style: ShellStyle) :
    DrumSetInstrument(context, hits) {
    /** The drum. */
    private val drum =
        context.modelD(style.tomModel, style.shellTexture).apply {
            recoilNode.attachChild(this)
            localScale = Vector3f.UNIT_XYZ.clone().mult(pitch.scale)
        }

    /** The stick. */
    private val stick: Striker =
        Striker(context, hits, StickType.DRUM_SET_STICK).apply {
            setParent(recoilNode)
        }

    init {
        // Move and rotate tom based on its pitch
        geometry.localTranslation = pitch.location()
        geometry.localRotation = pitch.rotation()

        // Move stick so that it strikes on the drum
        stick.node.move(STICK_NODE_OFFSET)
    }

    override fun tick(
        time: Duration,
        delta: Duration,
    ) {
        super.tick(time, delta)
        val result = stick.tick(time, delta)
        recoilDrum(
            drum = recoilNode,
            velocity = result.strike?.velocity ?: 0,
            delta = delta,
        )
    }
}

/** The pitch of the tom. */
@Suppress("kotlin:S6218", "ArrayInDataClass")
@Serializable
data class TomPitch(
    /**
     * The name of the pitch.
     */
    val name: String,
    /**
     * The scale of the pitch.
     */
    val scale: Float,
    private val location: Array<Float>,
    private val rotation: Array<Float>,
) {
    /** Returns the location of this as a [Vector3f]. */
    fun location(): Vector3f {
        return Vector3f(location[0], location[1], location[2])
    }

    /** Returns the rotation of this as a [Quaternion]. */
    fun rotation(): Quaternion {
        return Quaternion().fromAngles(
            rad(rotation[0]),
            rad(rotation[1]),
            rad(rotation[2]),
        )
    }

    companion object {
        private val pitches =
            Json.decodeFromString<Array<TomPitch>>(Utils.resourceToString("/instrument/alignment/Tom.json"))

        /** Retrieves a [TomPitch] based on its name from the resource file. */
        operator fun get(name: String): TomPitch = pitches.first { it.name == name }
    }
}
